extensions/ONNX/Onnx/OnnxTextGenerator.cs

```diff
@@ -0,0 +1,178 @@
+
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<string> GenerateTextAsync(
+        string prompt,
+        TextGenerationOptions? options = null,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        var tokens = this._tokenizer?.Encode(prompt);
+        using var generatorParams = new GeneratorParams(this._model);
+
+        if (options != null)
+        {
+            if (options.NucleusSampling > 0 && options.NucleusSampling <= 1)
+            {
+                this._config.NucleusSampling = options.NucleusSampling;
+            }
+            if (options.MaxTokens > 0)
+            {
+                this.MaxTokenTotal = (int)options.MaxTokens;
+            }
+            this._config.ResultsPerPrompt = options.ResultsPerPrompt;
+            this.Temperature = options.Temperature;
+        }
+
+        generatorParams.SetSearchOption("max_length", this.MaxTokenTotal);
+        generatorParams.SetSearchOption("min_length", this._config.MinLength);
+        generatorParams.SetSearchOption("num_return_sequences", this._config.ResultsPerPrompt);
+        generatorParams.SetSearchOption("temperature", this.Temperature);
+        generatorParams.SetSearchOption("repetition_penalty", this._config.RepetitionPenalty);
+        generatorParams.SetSearchOption("length_penalty", this._config.LengthPenalty);
+
+        switch (this._config.SearchType)
+        {
+            case OnnxSearchType.BeamSearch:
+                generatorParams.SetSearchOption("do_sample", false);
+                generatorParams.SetSearchOption("early_stopping", this._config.EarlyStopping);
+
+                if (this._config.NumBeams != null)
+                {
+                    generatorParams.SetSearchOption("num_beams", (double)this._config.NumBeams);
+                }
+                break;
+            case OnnxSearchType.TopN:
+                generatorParams.SetSearchOption("do_sample", true);
+                generatorParams.SetSearchOption("top_k", this._config.TopK);
+                generatorParams.SetSearchOption("top_p", this._config.NucleusSampling);
+                break;
+            default:
+
+                generatorParams.SetSearchOption("do_sample", false);
+
+                if (this._config.NumBeams != null)
+                {
+                    generatorParams.SetSearchOption("num_beams", (double)this._config.NumBeams);
+                }
+                break;
+        }
+
+        generatorParams.SetInputSequences(tokens);
+
+        using (var generator = new Generator(this._model, generatorParams))
+        {
+            List<int> outputTokens = new();
+
+            while (!generator.IsDone() && cancellationToken.IsCancellationRequested == false)
+            {
+                generator.ComputeLogits();
+                generator.GenerateNextToken();
+
+                await Task.Run(() => outputTokens.AddRange(generator.GetSequence(0)), cancellationToken).ConfigureAwait(true);
```

Comment:
I would suggest using a `await Task.CompletedTask.ConfigureAwait(false);` at the end of the method instead, to avoid using multiple tasks here.

Reply:
Thats a clever way to implement an async :)